### ISSUE TITLE
"Scanning:" Ins-Completion Messages

### ISSUE THREAD

**Issue by dylnmc** [vim/vim#3412]
I didn't see this in `'shortmess'` options, so I am posting this here so that maybe someone can help me or empathize and perhaps submit a patch (I have never done that and don't know how vim works at all--sorry).

Every time I press `i_CTRL-N` I get all of the messages that you should see below in the code snippet--one for every file. It is causing a lot of clutter. I don't necessarily rely on `:messages`, but I do use it perhaps more than most people (sometimes I refer to it for fixed lint errors and sometimes for debugging small vimscripts which are actually a significant routine in my work life, since I have found it necessary to automate many things and customize the crap out of certain things).

```
(1 of 2): Command line completion: Scanning for tags doesn't check for typed key now and
Scanning: js/banana.js
Scanning: js/foo.js
Scanning: js/bar.js
Scanning: foo.html
Scanning: data/data.txt
Scanning: ../index.html
Scanning: styles/foo.css
Scanning tags.
1 line less; before #2269  2 seconds ago
```

I would sincerely appreciate if someone could fix this or suggest a fix. Perhaps another option in `'shortmess'` would be helpful. I even tried `inoremap <silent> <c-n> <c-n>` and `inoremap <silent> <c-p> <c-p>` and it didn't work.

**Comment by tonymec** [comment]
What is the answer to
```vim
:verbose set verbose?
```
?

The higher the value, well, the more verbose Vim becomes. For example, if `'verbose'` is set to 5 or higher, Vim reports every searched tags and include file. My preferred setting is 1 (which answers were options, mappings, highlights and autocommands were set when queried about their values) and the default is zero.

Best regards,
Tony.

**Comment by dylnmc** [comment]
`verbose=1`

Still seems to happen with `set verbose=0` ... which doesn't seem to be able to change `:verbose set verbose?`

btw version:

```
VIM - Vi IMproved 8.1 (2018 May 18, compiled Aug 14 2018 22:10:17)
MS-Windows 32-bit GUI version with OLE support
Included patches: 1-287
Compiled by appveyor@APPVYR-WIN
Huge version with GUI.  Features included (+) or not (-):
+acl                +clipboard          +directx            +gettext/dyn        +localmap           +packages           +scrollbind         -tgetent            +vreplace
+arabic             +cmdline_compl      -dnd                -hangul_input       +lua/dyn            +path_extra         +signs              -termresponse       -vtp
+autocmd            +cmdline_hist       -ebcdic             +iconv/dyn          +menu               +perl/dyn           +smartindent        +textobjects        +wildignore
+autochdir          +cmdline_info       +emacs_tags         +insert_expand      +mksession          +persistent_undo    +startuptime        +timers             +wildmenu
+autoservername     +comments           +eval               +job                +modify_fname       -postscript         +statusline         +title              +windows
+balloon_eval       +conceal            +ex_extra           +jumplist           +mouse              +printer            -sun_workshop       +toolbar            +writebackup
-balloon_eval_term  +cryptv             +extra_search       +keymap             +mouseshape         +profile            +syntax             +user_commands      -xfontset
+browse             +cscope             +farsi              +lambda             +multi_byte_ime/dyn +python/dyn         +tag_binary         +vartabs            -xim
++builtin_terms     +cursorbind         +file_in_path       +langmap            +multi_lang         +python3/dyn        +tag_old_static     +vertsplit          +xpm_w32
+byte_offset        +cursorshape        +find_in_path       +libcall            +mzscheme/dyn       +quickfix           -tag_any_white      +virtualedit        -xterm_save
+channel            +dialog_con_gui     +float              +linebreak          +netbeans_intg      +reltime            +tcl/dyn            +visual             
+cindent            +diff               +folding            +lispindent         +num64              +rightleft          -termguicolors      +visualextra        
+clientserver       +digraphs           -footer             +listcmds           +ole                +ruby/dyn           +terminal           +viminfo            
   system vimrc file: "$VIM\vimrc"
     user vimrc file: "$HOME\_vimrc"
 2nd user vimrc file: "$HOME\vimfiles\vimrc"
 3rd user vimrc file: "$VIM\_vimrc"
      user exrc file: "$HOME\_exrc"
  2nd user exrc file: "$VIM\_exrc"
  system gvimrc file: "$VIM\gvimrc"
    user gvimrc file: "$HOME\_gvimrc"
2nd user gvimrc file: "$HOME\vimfiles\gvimrc"
3rd user gvimrc file: "$VIM\_gvimrc"
       defaults file: "$VIMRUNTIME\defaults.vim"
    system menu file: "$VIMRUNTIME\menu.vim"
Compilation: cl -c /W3 /nologo  -I. -Iproto -DHAVE_PATHDEF -DWIN32  -DFEAT_CSCOPE -DFEAT_TERMINAL -DFEAT_NETBEANS_INTG -DFEAT_JOB_CHANNEL   -DFEAT_XPM_W32   -DWINVER=0x0501 -D_WIN32_WINNT=0x0501 /MP -DHAVE_STDINT_H /Ox /GL -DNDEBUG /arch:IA32 /Zl /MT -DFEAT_OLE -DFEAT_MBYTE_IME -DDYNAMIC_IME -DFEAT_MBYTE -DFEAT_GUI_W32 -DFEAT_DIRECTX -DDYNAMIC_DIRECTX -DFEAT_DIRECTX_COLOR_EMOJI -DDYNAMIC_ICONV -DDYNAMIC_GETTEXT -DFEAT_TCL -DDYNAMIC_TCL -DDYNAMIC_TCL_DLL=\"tcl86t.dll\" -DDYNAMIC_TCL_VER=\"8.6\" -DFEAT_LUA -DDYNAMIC_LUA -DDYNAMIC_LUA_DLL=\"lua53.dll\" -DFEAT_PYTHON -DDYNAMIC_PYTHON -DDYNAMIC_PYTHON_DLL=\"python27.dll\" -DFEAT_PYTHON3 -DDYNAMIC_PYTHON3 -DDYNAMIC_PYTHON3_DLL=\"python36.dll\" -DFEAT_MZSCHEME -I "C:\Program Files (x86)\Racket\include" -DMZ_PRECISE_GC -DDYNAMIC_MZSCHEME -DDYNAMIC_MZSCH_DLL=\"libracket3m_a36fs8.dll\" -DDYNAMIC_MZGC_DLL=\"libracket3m_a36fs8.dll\" -DFEAT_PERL -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DDYNAMIC_PERL -DDYNAMIC_PERL_DLL=\"perl524.dll\" -DFEAT_RUBY -DDYNAMIC_RUBY -DDYNAMIC_RUBY_VER=24 -DDYNAMIC_RUBY_DLL=\"msvcrt-ruby240.dll\" -DFEAT_HUGE /Fd.\ObjGXOULYHTRZi386/ /Zi
Linking: link  /nologo /subsystem:windows,5.01 /opt:ref /LTCG:STATUS oldnames.lib kernel32.lib advapi32.lib shell32.lib gdi32.lib  comdlg32.lib ole32.lib netapi32.lib uuid.lib /machine:i386 gdi32.lib version.lib   winspool.lib comctl32.lib advapi32.lib shell32.lib netapi32.lib  /machine:i386  libcmt.lib oleaut32.lib user32.lib  /nodefaultlib:lua53.lib  /STACK:8388608  /nodefaultlib:python27.lib /nodefaultlib:python36.lib   "C:\Tcl\lib\tclstub86.lib" WSock32.lib xpm\x86\lib-vc14\libXpm.lib /PDB:gvim.pdb -debug
```

---

@tonymec

**Comment by chrisbra** [comment]
So what does `:verbose :set verbose?` return? Also please try reproducing using `vim --clean`.

BTW: I think this is a configuration issue and no Vim bug, so I am closing this. Please use the vim mailinglists or vi.stackexchange.com to get assistance in how to debug your vim.

**Comment by dylnmc** [comment]
I told you that `:verbose set verbose?` returns `verbose=1`. If I run the command `:set verbose=0` then `:verbose set verbose?` still returns `verbose=1`.

It's not a configuration issue. I ran it with --clean. re-open or I will open a new issue. thanks for the links to help others debug vim issues

@chrisbra

**Comment by chrisbra** [comment]
`:verbose set verbose?` actually will temporarily set the verbose option so that Vim will output where it was last set. So the output of 1 is expected (sorry, my error here). However, try `:echo &verbose` to make sure it is set to zero.

If you still think that this is a bug of vim, please provide a reproducible exmaple starting with `vim --clean`. Because it doesn't look like that.

BTW: I don't really like your attitude of trying to force us doing here something. Remember, this is an open source project and we all do it in our spare time without being paid for.

Oh and BTW, you don't need to mention me here explicitly, because I get the notifications to the ticket anyhow.

**Comment by dylnmc** [comment]
you closed the issue because you assumed the issue without delving into it. I was nice at first.. I'm not forcing anyone to do anything, and I thought I had very clearly and nicely communicated the issue and how to reproduce it. Allow me to be even more clear:

1. `vim --clean -u NONE`<kbd>Return</kbd>
2. `:set nocompatible hidden`<kbd>Return</kbd>
3. `:e foo.bar`<kbd>Return</kbd>
4. `iHello from foo.bar`<kbd>Escape</kbd>
5. `:e bar.foo`<kbd>Return</kbd>
6. `iHelloooooo from bar.foo`<kbd>Escape</kbd>
7. `:e new`<kbd>Return</kbd>
8. `iHe`<kbd>ctrl</kbd>+<kbd>n</kbd><kbd>Escape</kbd>
9. `:redir @" | silent messages | redir end`<kbd>Return</kbd>
10. `p`

For me this yields a lot of unwarranted garble ... I don't want to force anyone to fix this. I was just hoping that someome would sympathize with me and submit a patch or have a way of silencing all of the nonsense.

`'verbose'` wasn't set from anywhere, so it doesn't have an origin as to where it was set. I said that it says `verbose=1` and that is all that it says.

Thanks,

---

@chrisbra

**Comment by dylnmc** [comment]
@chrisbra I just saw the :echo &verbose .. I hadn't tried that, but that *is* `0`.

I was never trying to be rude or push the .. whatever .. and make anyone do anything. I just didn't see how this was a config issue. I'm honestly too tired to care about this. If it gets fixed or someone has a fix | :confetti_ball: | Else | k | Endif

**Comment by k-takata** [comment]
Using `:silent` doesn't affect to `:redir`. `:help :silent` says that:

```
			Redirection, started with |:redir|, will continue as
			usual, although there might be small differences.
```

> For example, if 'verbose' is set to 5 or higher, Vim reports every searched tags and include file.

The source code related to tags in tag.c and misc2.c checks whether 'verbose' is 5 or higher, however `ins_compl_get_exp()` in edit.c doesn't check 'verbose'. So the `Scanning:` messages will be shown even if 'verbose' is 0.

**Comment by dylnmc** [comment]
If I do `redir @" | messages | redir end`, then I see a long list of messages. If I throw in the `silent` then it prevents that. I am not sure if your vim works the same. (Edit: I think I see what you were saying. I was just using redir to more effectively illustrate the spam in `:messages` I just like to have a more tactile version of `:messages`, so I often end up redirecting it to @+ or @")

That seems like the issue there. Thanks

@k-takata

**Comment by chrisbra** [comment]
So would that patch fix it?

```patch
diff --git a/src/edit.c b/src/edit.c
index ff45bb7ee..3d083dd5f 100644
--- a/src/edit.c
+++ b/src/edit.c
@@ -4494,13 +4494,18 @@ ins_compl_get_exp(pos_T *ini)
                    dict = ins_buf->b_fname;
                    dict_f = DICT_EXACT;
                }
-               vim_snprintf((char *)IObuff, IOSIZE, _("Scanning: %s"),
+               if (p_verbose >= 5)
+               {
+                   verbose_enter();
+                   vim_snprintf((char *)IObuff, IOSIZE, _("Scanning: %s"),
                        ins_buf->b_fname == NULL
                            ? buf_spname(ins_buf)
                            : ins_buf->b_sfname == NULL
                                ? ins_buf->b_fname
                                : ins_buf->b_sfname);
-               (void)msg_trunc_attr(IObuff, TRUE, HL_ATTR(HLF_R));
+                   (void)msg_trunc_attr(IObuff, TRUE, HL_ATTR(HLF_R));
+                   verbose_leave();
+               }
            }
            else if (*e_cpt == NUL)
                break;
```

**Comment by dylnmc** [comment]
I'm no c expert, and I've never had a look at vim src code. So, I suspect that you're asking other active members.

Looks good, and thanks for looking into it.

**Comment by k-takata** [comment]
I haven't tested the patch, but it might not be enough. It seems that the following messages are also shown:
https://github.com/vim/vim/blob/274cea35c6d409cfb16862ff24e789b2d3ea3ee3/src/edit.c#L4532
https://github.com/vim/vim/blob/274cea35c6d409cfb16862ff24e789b2d3ea3ee3/src/edit.c#L5113
And this is also a similar one? (Not sure how to reproduce.)
https://github.com/vim/vim/blob/274cea35c6d409cfb16862ff24e789b2d3ea3ee3/src/edit.c#L3395

I'm not sure whether it is safe to disable these messages by default. Is there any plugin that rely on the current behavior? (I hope there are no such plugins.)

**Comment by k-takata** [comment]
It seems that this was mistakenly closed. Reopening.

**Comment by chrisbra** [comment]
Yeah, sorry, I haven't had the time to fix the remaining problems yet. Not enough spare time for me and too many private and professional things to do... It is still on my list so I hope to come back to it sooner or later.

**Comment by lacygoill** [comment]
> And this is also a similar one? (Not sure how to reproduce.)

Yes, I think it's a similar message which is displayed when the dictionary completion is used (`:h i^x^k`). You can see it after starting Vim with this shell command:

    $ vim -Nu NONE +'set dict=/usr/share/dict/words' +startinsert +'call feedkeys("simul")'

Then pressing `C-x C-k` to complete `simul`.
The command may need to be tweaked because it assumes that there exists a dictionary file in `/usr/share/dict/words`.

After the first `C-x C-k`, this message is briefly displayed:

    Scanning dictionary: /usr/share/dict/words

But it's almost immediately erased and replaced by this message:

    -- Dictionary completion (^K^N^P) match 12 of 34

The actual message depends on whether `'showmode'` is set and whether `'shortmess'` contains the `c` flag:

    ┌────────────────┬────────────────────────────────────────────────────────────────────────────────────────┐
    │                │                                 shortmess contains 'c'                                 │
    │                ├─────────────────────────────────────┬──────────────────────────────────────────────────┤
    │                │                 yes                 │                        no                        │
    ├──────────┬─────┼─────────────────────────────────────┼──────────────────────────────────────────────────┤
    │ showmode │ yes │          -- INSERT --               │ -- Dictionary completion (^K^N^P) match 12 of 34 │
    │    is    ├─────┼─────────────────────────────────────┼──────────────────────────────────────────────────┤
    │    on    │ no  │ match in file /usr/share/dict/words │                  match 12 of 34                  │
    └──────────┴─────┴─────────────────────────────────────┴──────────────────────────────────────────────────┘

However, the message is logged no matter what in the output of `:messages`.
After that, every time you select a different match in the pum, an additional message is logged.

I frequently use the dictionary completion, so I regularly see these messages in the output of `:messages`, but also on the command-line because I reset `'showmode'`.

**Comment by nhooyr** [comment]
Can confirm that setting `showmode` and including `c` in `shortmess` removes these messages.

**Comment by nhooyr** [comment]
As a workaround for now I've set the following in my colorscheme to make the mode msg disappear:

```
hi ModeMsg      ctermfg=white ctermbg=NONE cterm=NONE
```

This way I don't see the mode message or the scanning messages.

**Comment by nhooyr** [comment]
Unfortunately, I still see `Scanning` messages occasionally flash but then disappear.

**Comment by nhooyr** [comment]
I just disabled tag completion with `set complete-=t`

**Comment by mg979** [comment]
By the way nvim doesn't have this problem, that is: it prints the `Scanning...` message, but at least it doesn't show up it in the `:messages` output. I don't think there's a good reason the `:messages` output should be spammed with these.

Maybe just fix it the nvim way? It's annoying since it makes `:messages` nearly useless.

My `&verbose` setting is `0`.

**Comment by lacygoill** [comment]
Not sure, but possibly fixed by [8.2.2130](https://github.com/vim/vim/releases/tag/v8.2.2130).

**Comment by brammool** [comment]
Yes, this appears to be fixed now.
